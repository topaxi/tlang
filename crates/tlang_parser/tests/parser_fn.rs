use indoc::indoc;

mod common;

#[test]
fn test_function_declaration() {
    assert_parser_snapshot!("fn foo() { bar(); 1 + 2; }");
}

#[test]
fn test_function_declaration_with_parameters() {
    assert_parses!("fn foo(x, y) { 1 + 2; }");
}

#[test]
fn test_nameless_function_expressions() {
    assert_parses!("let x = fn() { 1 + 2; };");
    assert_parses!("let x = fn(x, y) { 1 + 2; };");
}

#[test]
fn test_function_expression_without_name_no_argument_parenthesis() {
    assert_parses!("let x = fn { 1 + 2; };");
}

#[test]
fn test_function_expressions() {
    assert_parses!("let x = fn foo() { 1 + 2; };");
    assert_parses!("let x = fn foo(x, y) { 1 + 2; };");
}

#[test]
fn test_explicit_return_statements() {
    assert_parses!("fn foo() { return 1 + 2; }");
    assert_parses!("fn foo() { return; }");
    assert_parses!("let x = fn() { return 1 + 2; };");
}

#[test]
fn test_implicit_return_expressions() {
    assert_parses!("fn foo() { 1 + 2 }");
    assert_parses!("let x = fn() { 1 + 2 };");
    assert_parser_snapshot!("fn foo() { let x = 1; x }");
}

#[test]
fn test_recursive_factorial_functional_definition() {
    assert_parser_snapshot!(indoc! {"
        fn factorial(0) { return 1; }
        fn factorial(n) { return n * factorial(n - 1); }
    "});
}

#[test]
fn test_recursive_sum() {
    assert_parser_snapshot!(indoc! {"
        fn sum([]) { 0 }
        fn sum([x, ...xs]) { x + sum(xs) }
    "});
}

#[test]
fn test_functional_function_declaration_with_comments_inbetween() {
    assert_parser_snapshot!(indoc! {"
        fn foo(1) { 1 }
        // comment
        fn foo(n) { n * 2 }
    "});
}

#[test]
fn test_explicit_tail_recursive_call() {
    assert_parser_snapshot!(indoc! {"
        fn factorial(n) { factorial(n, 1) }
        fn factorial(0, acc) { acc }
        fn factorial(n, acc) { rec factorial(n - 1, n * acc) }
    "});
}

#[test]
fn test_foldl_impl() {
    assert_parser_snapshot!(indoc! {"
        fn foldl([], acc, _) { acc }
        fn foldl([x, ...xs], acc, f) { rec foldl(xs, f(acc, x), f) }
    "});
}

#[test]
fn test_function_declarations_with_guard() {
    assert_parser_snapshot!(indoc! {"
        fn filter([], f) { [] }
        fn filter([x, ...xs], f) if f(x) { [x, ...filter(xs, f)] }
        fn filter([x, ...xs], f) { rec filter(xs, f) }
    "});
}

#[test]
fn test_function_declarations_with_let_guard() {
    assert_parser_snapshot!(indoc! {"
        fn filter_map([], f) { [] }
        fn filter_map([x, ...xs], f) if let y = f(x) { [y, ...filter_map(xs, f)] }
        fn filter_map([x, ...xs], f) { rec filter_map(xs, f) }
    "});
}

#[test]
#[should_panic]
fn test_parameter_list_mandatory_comma() {
    parse!("fn foo(x y) {}");
}

#[test]
fn test_list_matching_wildcard() {
    assert_parser_snapshot!("fn tail([_, ...xs]) { xs }");
}

#[test]
fn test_fn_expression_in_function_completion_position() {
    assert_parser_snapshot!("fn foo() { fn bar() {} }");
}
