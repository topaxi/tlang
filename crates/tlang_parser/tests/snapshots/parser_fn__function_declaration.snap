---
source: crates/tlang_parser/tests/parser_fn.rs
expression: "parse!(\"fn foo() { bar(); 1 + 2; }\")"
---
Node(
  ast_node: Legacy(Module([
    Node(
      ast_node: Legacy(FunctionSingleDeclaration(
        id: SymbolId(1),
        name: Node(
          ast_node: Legacy(Identifier("foo")),
          symbol_table: None,
          span: Span(
            start: LineColumn(
              line: 0,
              column: 3,
            ),
            end: LineColumn(
              line: 0,
              column: 6,
            ),
          ),
        ),
        declaration: Node(
          ast_node: Legacy(FunctionDeclaration(FunctionDeclaration(
            parameters: [],
            guard: None,
            return_type_annotation: None,
            body: Node(
              ast_node: Legacy(Block([
                Node(
                  ast_node: Legacy(ExpressionStatement(Node(
                    ast_node: Legacy(Call(
                      function: Node(
                        ast_node: Legacy(Identifier("bar")),
                        symbol_table: None,
                        span: Span(
                          start: LineColumn(
                            line: 0,
                            column: 11,
                          ),
                          end: LineColumn(
                            line: 0,
                            column: 14,
                          ),
                        ),
                      ),
                      arguments: [],
                    )),
                    symbol_table: None,
                    span: Span(
                      start: LineColumn(
                        line: 0,
                        column: 11,
                      ),
                      end: LineColumn(
                        line: 0,
                        column: 16,
                      ),
                    ),
                  ))),
                  symbol_table: None,
                  span: Span(
                    start: LineColumn(
                      line: 0,
                      column: 11,
                    ),
                    end: LineColumn(
                      line: 0,
                      column: 17,
                    ),
                  ),
                ),
                Node(
                  ast_node: Legacy(ExpressionStatement(Node(
                    ast_node: Legacy(BinaryOp(
                      op: Add,
                      lhs: Node(
                        ast_node: Legacy(Literal(Integer(1))),
                        symbol_table: None,
                        span: Span(
                          start: LineColumn(
                            line: 0,
                            column: 18,
                          ),
                          end: LineColumn(
                            line: 0,
                            column: 19,
                          ),
                        ),
                      ),
                      rhs: Node(
                        ast_node: Legacy(Literal(Integer(2))),
                        symbol_table: None,
                        span: Span(
                          start: LineColumn(
                            line: 0,
                            column: 22,
                          ),
                          end: LineColumn(
                            line: 0,
                            column: 23,
                          ),
                        ),
                      ),
                    )),
                    symbol_table: None,
                    span: Span(
                      start: LineColumn(
                        line: 0,
                        column: 18,
                      ),
                      end: LineColumn(
                        line: 0,
                        column: 23,
                      ),
                    ),
                  ))),
                  symbol_table: None,
                  span: Span(
                    start: LineColumn(
                      line: 0,
                      column: 18,
                    ),
                    end: LineColumn(
                      line: 0,
                      column: 24,
                    ),
                  ),
                ),
              ], None)),
              symbol_table: None,
              span: Span(
                start: LineColumn(
                  line: 0,
                  column: 0,
                ),
                end: LineColumn(
                  line: 0,
                  column: 0,
                ),
              ),
            ),
          ))),
          symbol_table: None,
          span: Span(
            start: LineColumn(
              line: 0,
              column: 0,
            ),
            end: LineColumn(
              line: 0,
              column: 0,
            ),
          ),
        ),
      )),
      symbol_table: None,
      span: Span(
        start: LineColumn(
          line: 0,
          column: 0,
        ),
        end: LineColumn(
          line: 0,
          column: 26,
        ),
      ),
    ),
  ])),
  symbol_table: None,
  span: Span(
    start: LineColumn(
      line: 0,
      column: 0,
    ),
    end: LineColumn(
      line: 0,
      column: 0,
    ),
  ),
)
